amr-wind/wind_energy/actuator/disk/constant_ct_ops.H

```diff
@@ -0,0 +1,167 @@
+{
+    void operator()(ConstantCt::DataType& data, const utils::ActParser& pp)
+    {
+        // TODO get density at probe location
+        check_for_parse_conflicts(pp);
+        auto& meta = data.meta();
+        required_parameters(meta, pp);
+        optional_parameters(meta, pp);
+        final_checks(meta);
+        do_parse_based_computations(data);
+    }
+};
+
+template <>
+struct InitDataOp<ConstantCt, ActSrcDisk>
+{
+    void operator()(ConstantCt::DataType& data)
+    {
+        auto& grid = data.grid();
+        auto& meta = data.meta();
+
+        // only resize the members we are going to use
+        grid.pos.resize(meta.num_force_pts);
+        grid.force.resize(meta.num_force_pts);
+        grid.vel.resize(meta.num_vel_pts);
+        grid.vel_pos.resize(meta.num_vel_pts);
+
+        const auto& cVec = meta.coplanar_vec;
+        const auto& sVec = meta.sample_vec;
+
+        const auto& cc = meta.center;
+        {
+            const auto& dr = meta.dr;
+            for (int i = 0; i < meta.num_force_pts; ++i) {
+                grid.pos[i] = cc + (i + 0.5) * dr * cVec;
+            }
+        }
+        {
+            const auto rotVec = vs::Vector::khat() ^ sVec;
+            // vectors are normalized so magnitude of each is 1
+            const amrex::Real angle = std::acos((sVec & vs::Vector::ihat()));
+            const auto rotMatrix = vs::quaternion(rotVec, angle);
+
+            const vs::Vector nvp = {meta.num_vel_pts_r, meta.num_vel_pts_t, 1};
+
+            const amrex::Real dr = meta.diameter / nvp.x();
+            const amrex::Real dt = ::amr_wind::utils::two_pi() / nvp.y();
+            const amrex::Real du = meta.diameters_to_sample * meta.diameter;
+
+            int ip = 0;
+            for (int i = 0; i < nvp.x(); i++) {
+                const amrex::Real r = dr * (i + 0.5);
+                for (int j = 0; j < nvp.y(); j++, ip++) {
+                    const amrex::Real theta = j * dt;
+                    vs::Vector refPoint = {
+                        r * std::cos(theta), r * std::sin(theta), du};
+                    grid.vel_pos[ip] = (refPoint & rotMatrix) + cc;
+                }
+            }
+        }
+    }
+};
+
+template <>
+struct UpdateVelOp<ConstantCt, ActSrcDisk>
+{
+    void operator()(ConstantCt::DataType& data)
+    {
+        auto& meta = data.meta();
+        const auto& grid = data.grid();
+        auto& refVel = meta.reference_velocity;
+        for (auto&& v : grid.vel) {
+            refVel = refVel + v;
+        }
+        refVel /= grid.vel.size();
+    }
+};
+
+// compute the total force over a given radial section
+// this will then be spread uniformly over that area in the
+// projection step
+template <>
+struct ComputeForceOp<ConstantCt, ActSrcDisk>
+{
+    void operator()(ConstantCt::DataType& data)
+    {
+        auto& grid = data.grid();
+        auto& ddata = data.meta();
+        const amrex::Real uInf = ddata.reference_velocity & ddata.normal_vec;
```

Comment:
@tonyinme no it can be set to be sampled upstream by setting the parameter `diameters_to_sample`.  Basically we sample a projected disk upstream (or downstream if for some reason you set this to a negative value) and average over all the points to get the reference velocity. Then `uInf` is taken as the disk normal component of this value.  

You can also sample in a direction other than the disk normal. So if the turbine is yawed you can sample without yaw to get the "free stream" and then this projection will make sure that only the disk normal component is applied when computing the force. I've update the regression test to show how this works.

Reply:
I should add that if you don't add the `diameters_to_sample` flag then it's value will default to 0 and sampling will be at the disk.